Loadtafsir

### Steps to reproduce

Idk

### Expected behavior

When i select tafsirs it should tafsirs appears 

### Actual behavior

It stays no tafsirs available also that happen with Reciter 

### QuranApp version

2023.09.03.1

### Android version

10

### Other details

java.net.UnknownHostException: Unable to resolve host "cdn.jsdelivr.net": No address associated with hostname
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:156)
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:103)
	at java.net.InetAddress.getAllByName(InetAddress.java:1152)
	at ea.p.b(SourceFile:218)
	at ha.e.a(SourceFile:131)
	at ha.e.b(SourceFile:1)
	at ga.a.a(SourceFile:441)
	at ia.f.b(SourceFile:143)
	at ia.f.a(Unknown Source:4)
	at ga.a.a(SourceFile:140)
	at ia.f.b(SourceFile:143)
	at ia.f.a(Unknown Source:4)
	at ia.a.a(SourceFile:247)
	at ia.f.b(SourceFile:143)
	at ia.a.a(SourceFile:506)
	at ia.f.b(SourceFile:143)
	at ia.f.a(Unknown Source:4)
	at o5.d.a(SourceFile:5)
	at ia.f.b(SourceFile:143)
	at ia.f.a(Unknown Source:4)
	at ea.z.d(SourceFile:87)
	at ea.y.a(SourceFile:19)
	at fa.a.run(SourceFile:18)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)
Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
	at libcore.io.Linux.android_getaddrinfo(Native Method)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:74)
	at libcore.io.BlockGuardOs.android_getaddrinfo(BlockGuardOs.java:200)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:74)
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:135)
	... 25 more


### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.

--- a/app/src/main/java/com/quranapp/android/QuranApp.kt
+++ b/app/src/main/java/com/quranapp/android/QuranApp.kt
@@ -5,6 +5,8 @@ import android.content.Context
 import android.os.Build
 import android.webkit.WebView
 import androidx.appcompat.app.AppCompatDelegate
+import com.quranapp.android.api.RetrofitInstance
+import com.quranapp.android.utils.app.DownloadSourceUtils
 import com.quranapp.android.utils.app.NotificationUtils
 import com.quranapp.android.utils.app.ThemeUtils
 import com.quranapp.android.utils.exceptions.CustomExceptionHandler
@@ -27,6 +29,7 @@ class QuranApp : Application() {
 
     override fun onCreate() {
         super.onCreate()
+        DownloadSourceUtils.resetDownloadSourceBaseUrl(this)
         NotificationUtils.createNotificationChannels(this)
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {

--- a/app/src/main/java/com/quranapp/android/activities/ActivityReadHistory.kt
+++ b/app/src/main/java/com/quranapp/android/activities/ActivityReadHistory.kt
@@ -4,7 +4,10 @@ import android.content.Intent
 import android.os.Bundle
 import android.view.View
 import android.view.ViewGroup
+import android.widget.Toast
 import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.ItemTouchHelper
+import androidx.recyclerview.widget.RecyclerView
 import com.peacedesign.android.utils.ColorUtils
 import com.peacedesign.android.utils.WindowUtils
 import com.quranapp.android.R
@@ -21,6 +24,7 @@ import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+
 
 class ActivityReadHistory : QuranMetaPossessingActivity(), ReadHistoryCallbacks {
     private lateinit var binding: ActivityBookmarkBinding
@@ -106,14 +110,14 @@ class ActivityReadHistory : QuranMetaPossessingActivity(), ReadHistoryCallbacks 
             val models = readHistoryDBHelper?.getAllHistories(-1) ?: ArrayList()
 
             withContext(Dispatchers.Main) {
-                setupAdapter(models)
+                setupAdapter(ArrayList(models))
 
                 binding.loader.visibility = View.GONE
             }
         }
     }
 
-    private fun setupAdapter(models: List<ReadHistoryModel>) {
+    private fun setupAdapter(models: ArrayList<ReadHistoryModel>) {
         if (adapter == null && models.isNotEmpty()) {
             val spanCount = if (WindowUtils.isLandscapeMode(this)) 2 else 1
             binding.list.layoutManager = GridLayoutManager(this, spanCount)
@@ -127,6 +131,32 @@ class ActivityReadHistory : QuranMetaPossessingActivity(), ReadHistoryCallbacks 
 
         binding.header.setShowRightIcon(models.isNotEmpty())
         binding.noItems.visibility = if (models.isEmpty()) View.VISIBLE else View.GONE
+
+        val touchHelper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
+            override fun onMove(
+                recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder
+            ): Boolean {
+                return false
+            }
+
+            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
+                val position = viewHolder.bindingAdapterPosition
+
+                if (position < 0 || position >= models.size) {
+                    return
+                }
+
+                val model = models[position]
+                readHistoryDBHelper?.deleteHistory(model)
+                models.removeAt(position)
+                adapter?.notifyItemRemoved(position)
+
+                if (models.isEmpty()) {
+                    refreshHistories()
+                }
+            }
+        })
+        touchHelper.attachToRecyclerView(binding.list)
     }
 
     override fun onReadHistoryRemoved(model: ReadHistoryModel) {}

--- a/app/src/main/java/com/quranapp/android/api/ApiConfig.kt
+++ b/app/src/main/java/com/quranapp/android/api/ApiConfig.kt
@@ -7,7 +7,8 @@
 package com.quranapp.android.api
 
 object ApiConfig {
-    const val GITHUB_ROOT_URL = "https://cdn.jsdelivr.net/gh/AlfaazPlus/QuranApp@latest/"
+    const val JS_DELIVR_ROOT_URL = "https://cdn.jsdelivr.net/gh/AlfaazPlus/QuranApp@latest/"
+    const val GITHUB_ROOT_URL = "https://raw.githubusercontent.com/AlfaazPlus/QuranApp/master/"
     const val GITHUB_REPOSITORY_URL = "https://github.com/AlfaazPlus/QuranApp"
     const val GITHUB_ISSUES_BUG_REPORT_URL = "https://github.com/AlfaazPlus/QuranApp/issues/new?template=bug_report.yml"
     const val GITHUB_ISSUES_VERSE_REPORT_URL = "https://github.com/AlfaazPlus/QuranApp/issues/new?template=verse_report.yml"

--- a/app/src/main/java/com/quranapp/android/api/RetrofitInstance.kt
+++ b/app/src/main/java/com/quranapp/android/api/RetrofitInstance.kt
@@ -17,16 +17,24 @@ object RetrofitInstance {
         .cache(null)
         .build()
 
-    val github: GithubApi by lazy {
-        Retrofit.Builder()
-            .baseUrl(ApiConfig.GITHUB_ROOT_URL)
-            .addConverterFactory(
-                JsonHelper.json.asConverterFactory(MediaType.get("application/json"))
-            )
-            .client(client)
-            .build()
-            .create(GithubApi::class.java)
-    }
+    private var githubApi: GithubApi? = null
+    var githubResDownloadUrl: String = ApiConfig.JS_DELIVR_ROOT_URL
+
+    val github: GithubApi
+        get() {
+            if (githubApi == null) {
+                githubApi = Retrofit.Builder()
+                    .baseUrl(githubResDownloadUrl)
+                    .addConverterFactory(
+                        JsonHelper.json.asConverterFactory(MediaType.get("application/json"))
+                    )
+                    .client(client)
+                    .build()
+                    .create(GithubApi::class.java)
+            }
+
+            return githubApi!!
+        }
 
     val quran: QuranApi by lazy {
         Retrofit.Builder()
@@ -37,5 +45,9 @@ object RetrofitInstance {
             .client(client)
             .build()
             .create(QuranApi::class.java)
+    }
+
+    fun resetGithubApi() {
+        githubApi = null
     }
 }

--- a/app/src/main/java/com/quranapp/android/db/readHistory/ReadHistoryDBHelper.java
+++ b/app/src/main/java/com/quranapp/android/db/readHistory/ReadHistoryDBHelper.java
@@ -106,6 +106,19 @@ public class ReadHistoryDBHelper extends SQLiteOpenHelper {
         db.delete(TABLE_NAME, where, whereArgs);
     }
 
+
+    public void deleteHistory(@NonNull ReadHistoryModel model) {
+        deleteHistory(
+            null,
+            model.getReadType(),
+            model.getReaderStyle(),
+            model.getJuzNo(),
+            model.getChapterNo(),
+            model.getFromVerseNo(),
+            model.getToVerseNo()
+        );
+    }
+
     public void addToHistory(
         int readType, int readStyle, int juzNo, int chapterNo, int fromVerse, int toVerse,
         OnResultReadyCallback<ReadHistoryModel> callback

--- a/app/src/main/java/com/quranapp/android/frags/settings/FragSettingsMain.java
+++ b/app/src/main/java/com/quranapp/android/frags/settings/FragSettingsMain.java
@@ -6,26 +6,6 @@
 
 package com.quranapp.android.frags.settings;
 
-import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
-import static android.view.View.GONE;
-import static android.view.View.VISIBLE;
-import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
-import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
-import static android.widget.LinearLayout.HORIZONTAL;
-import static android.widget.LinearLayout.VERTICAL;
-import static com.quranapp.android.activities.readerSettings.ActivitySettings.KEY_SETTINGS_DESTINATION;
-import static com.quranapp.android.activities.readerSettings.ActivitySettings.SETTINGS_THEME;
-import static com.quranapp.android.activities.readerSettings.ActivitySettings.SETTINGS_VOTD;
-import static com.quranapp.android.reader_managers.ReaderParams.READER_READ_TYPE_VERSES;
-import static com.quranapp.android.reader_managers.ReaderParams.READER_STYLE_PAGE;
-import static com.quranapp.android.reader_managers.ReaderParams.READER_STYLE_TRANSLATION;
-import static com.quranapp.android.utils.reader.ReaderTextSizeUtils.TEXT_SIZE_MIN_PROGRESS;
-import static com.quranapp.android.utils.reader.ReaderTextSizeUtils.getMaxProgress;
-import static com.quranapp.android.utils.univ.Codes.SETTINGS_LAUNCHER_RESULT_CODE;
-import static com.quranapp.android.utils.univ.Keys.READER_KEY_SAVE_TRANSL_CHANGES;
-import static com.quranapp.android.utils.univ.Keys.READER_KEY_SETTING_IS_FROM_READER;
-import static com.quranapp.android.utils.univ.Keys.READER_KEY_TRANSL_SLUGS;
-
 import android.annotation.SuppressLint;
 import android.app.AlarmManager;
 import android.content.Context;
@@ -48,7 +28,6 @@ import android.view.View;
 import android.widget.LinearLayout;
 import android.widget.SeekBar;
 import android.widget.Toast;
-
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.annotation.RequiresApi;
@@ -58,6 +37,25 @@ import androidx.cardview.widget.CardView;
 import androidx.core.app.NotificationManagerCompat;
 import androidx.core.content.ContextCompat;
 import androidx.fragment.app.FragmentResultListener;
+import static com.quranapp.android.reader_managers.ReaderParams.READER_READ_TYPE_VERSES;
+import static com.quranapp.android.reader_managers.ReaderParams.READER_STYLE_PAGE;
+import static com.quranapp.android.reader_managers.ReaderParams.READER_STYLE_TRANSLATION;
+import static com.quranapp.android.utils.app.DownloadSourceUtils.DOWNLOAD_SRC_DEFAULT;
+import static com.quranapp.android.utils.app.DownloadSourceUtils.DOWNLOAD_SRC_GITHUB;
+import static com.quranapp.android.utils.app.DownloadSourceUtils.DOWNLOAD_SRC_JSDELIVR;
+import static com.quranapp.android.utils.reader.ReaderTextSizeUtils.TEXT_SIZE_MIN_PROGRESS;
+import static com.quranapp.android.utils.reader.ReaderTextSizeUtils.getMaxProgress;
+import static com.quranapp.android.utils.univ.Codes.SETTINGS_LAUNCHER_RESULT_CODE;
+import static com.quranapp.android.utils.univ.Keys.READER_KEY_SAVE_TRANSL_CHANGES;
+import static com.quranapp.android.utils.univ.Keys.READER_KEY_SETTING_IS_FROM_READER;
+import static com.quranapp.android.utils.univ.Keys.READER_KEY_TRANSL_SLUGS;
+import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
+import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
+import static android.widget.LinearLayout.HORIZONTAL;
+import static android.widget.LinearLayout.VERTICAL;
 
 import com.google.android.material.tabs.TabLayout;
 import com.peacedesign.android.utils.DrawableUtils;
@@ -65,6 +63,7 @@ import com.peacedesign.android.utils.WindowUtils;
 import com.peacedesign.android.widget.dialog.base.PeaceDialog;
 import com.quranapp.android.R;
 import com.quranapp.android.activities.readerSettings.ActivitySettings;
+import com.quranapp.android.api.RetrofitInstance;
 import com.quranapp.android.api.models.recitation.RecitationInfoModel;
 import com.quranapp.android.api.models.recitation.RecitationTranslationInfoModel;
 import com.quranapp.android.api.models.tafsir.TafsirInfoModel;
@@ -73,6 +72,7 @@ import com.quranapp.android.databinding.LytReaderIndexTabBinding;
 import com.quranapp.android.databinding.LytReaderSettingsItemBinding;
 import com.quranapp.android.databinding.LytReaderSettingsTextDecoratorBinding;
 import com.quranapp.android.databinding.LytReaderSettingsTextSizeBinding;
+import com.quranapp.android.databinding.LytResDownloadSourceSheetBinding;
 import com.quranapp.android.databinding.LytSettingsLayoutStyleBinding;
 import com.quranapp.android.databinding.LytSettingsReaderBinding;
 import com.quranapp.android.databinding.LytSettingsVotdToggleBinding;
@@ -81,6 +81,7 @@ import com.quranapp.android.frags.settings.appLogs.FragSettingsAppLogs;
 import com.quranapp.android.frags.settings.recitations.FragSettingsRecitations;
 import com.quranapp.android.frags.settings.recitations.manage.FragSettingsManageAudio;
 import com.quranapp.android.reader_managers.ReaderVerseDecorator;
+import com.quranapp.android.utils.app.DownloadSourceUtils;
 import com.quranapp.android.utils.app.ThemeUtils;
 import com.quranapp.android.utils.extensions.ContextKt;
 import com.quranapp.android.utils.extensions.LayoutParamsKt;
@@ -119,8 +120,6 @@ import kotlin.Unit;
 
 public class FragSettingsMain extends FragSettingsBase implements FragmentResultListener {
     private FragSettingsMainBinding mBinding;
-    private LytReaderSettingsItemBinding mThemeExplorerBinding;
-    private LytReaderSettingsItemBinding mVOTDToggleBinding;
     private LytReaderSettingsItemBinding mTranslExplorerBinding;
     private LytReaderSettingsItemBinding mTafsirExplorerBinding;
     private LytReaderSettingsItemBinding mRecitationExplorerBinding;
@@ -168,9 +167,6 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
         if (getContext() == null) {
             return;
         }
-
-        tryLaunchThemeExplorerFromIntent(getContext(), args);
-        tryLaunchVOTDToggleExplorer(getContext(), args);
     }
 
     @Override
@@ -277,31 +273,21 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
     }
 
     private void initThemes(LinearLayout parent) {
-        mThemeExplorerBinding = LytReaderSettingsItemBinding.inflate(mInflater);
-        setupLauncherParams(R.drawable.dr_icon_theme, mThemeExplorerBinding);
+        LytReaderSettingsItemBinding binding = LytReaderSettingsItemBinding.inflate(mInflater);
+        setupLauncherParams(R.drawable.dr_icon_theme, binding);
 
-        setupThemeTitle();
-        mThemeExplorerBinding.launcher.setOnClickListener(v -> launchThemeExplorer(v.getContext()));
+        setupThemeTitle(binding);
+        binding.launcher.setOnClickListener(v -> launchThemeExplorer(v.getContext(), binding));
 
-        parent.addView(mThemeExplorerBinding.getRoot());
-
-        tryLaunchThemeExplorerFromIntent(parent.getContext(), getArgs());
+        parent.addView(binding.getRoot());
     }
 
-    private void setupThemeTitle() {
-        if (mThemeExplorerBinding == null) return;
-
-        String selectedTheme = ThemeUtils.resolveThemeTextFromMode(mThemeExplorerBinding.getRoot().getContext());
-        prepareTitle(mThemeExplorerBinding, R.string.strTitleTheme, selectedTheme);
+    private void setupThemeTitle(LytReaderSettingsItemBinding binding) {
+        String selectedTheme = ThemeUtils.resolveThemeTextFromMode(binding.getRoot().getContext());
+        prepareTitle(binding, R.string.strTitleTheme, selectedTheme);
     }
 
-    private void tryLaunchThemeExplorerFromIntent(Context ctx, Bundle args) {
-        if (args.getInt(KEY_SETTINGS_DESTINATION) == SETTINGS_THEME) {
-            launchThemeExplorer(ctx);
-        }
-    }
-
-    private void launchThemeExplorer(Context ctx) {
+    private void launchThemeExplorer(Context ctx, LytReaderSettingsItemBinding parentBinding) {
         PeaceBottomSheet sheetDialog = new PeaceBottomSheet();
         PeaceBottomSheetParams params = sheetDialog.getParams();
         params.setHeaderTitle(getString(R.string.strTitleTheme));
@@ -332,45 +318,33 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
             return Unit.INSTANCE;
         });
 
-        sheetDialog.setOnDismissListener(this::setupThemeTitle);
+        sheetDialog.setOnDismissListener(() -> setupThemeTitle(parentBinding));
         sheetDialog.show(getParentFragmentManager());
     }
 
     private void initVOTDToggleLauncher(LinearLayout parent) {
-        mVOTDToggleBinding = LytReaderSettingsItemBinding.inflate(mInflater, parent, false);
+        LytReaderSettingsItemBinding binding = LytReaderSettingsItemBinding.inflate(mInflater, parent, false);
 
-        setupLauncherParams(R.drawable.dr_icon_heart_filled, mVOTDToggleBinding);
-        setupVOTDToggleTitle();
+        setupLauncherParams(R.drawable.dr_icon_heart_filled, binding);
+        setupVOTDToggleTitle(binding);
 
-        mVOTDToggleBinding.launcher.setOnClickListener(v -> launchVOTDToggleExplorer(v.getContext()));
+        binding.launcher.setOnClickListener(v -> launchVOTDToggleExplorer(v.getContext(), binding));
 
-        parent.addView(mVOTDToggleBinding.getRoot());
-
-        tryLaunchVOTDToggleExplorer(parent.getContext(), getArgs());
+        parent.addView(binding.getRoot());
     }
 
-    private void setupVOTDToggleTitle() {
-        if (mVOTDToggleBinding == null) {
-            return;
-        }
-
-        Context context = mVOTDToggleBinding.getRoot().getContext();
+    private void setupVOTDToggleTitle(LytReaderSettingsItemBinding binding) {
+        Context context = binding.getRoot().getContext();
 
         String status = context.getString(
             VOTDUtils.isVOTDTrulyEnabled(context)
                 ? R.string.strLabelOn
                 : R.string.strLabelOff
         );
-        prepareTitle(mVOTDToggleBinding, R.string.strTitleVOTD, status);
+        prepareTitle(binding, R.string.strTitleVOTD, status);
     }
 
-    private void tryLaunchVOTDToggleExplorer(Context ctx, Bundle args) {
-        if (args.getInt(KEY_SETTINGS_DESTINATION) == SETTINGS_VOTD) {
-            launchVOTDToggleExplorer(ctx);
-        }
-    }
-
-    private void launchVOTDToggleExplorer(Context ctx) {
+    private void launchVOTDToggleExplorer(Context ctx, LytReaderSettingsItemBinding parentBinding) {
         AlarmManager alarmManager = ContextCompat.getSystemService(ctx, AlarmManager.class);
 
         PeaceBottomSheet sheetDialog = new PeaceBottomSheet();
@@ -415,7 +389,7 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
             return true;
         });
 
-        sheetDialog.setOnDismissListener(this::setupVOTDToggleTitle);
+        sheetDialog.setOnDismissListener(() -> setupVOTDToggleTitle(parentBinding));
         sheetDialog.show(getParentFragmentManager());
     }
 
@@ -721,7 +695,8 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
             public void onStopTrackingTouch(@NonNull SeekBar seekBar) {
                 SPReader.setSavedTextSizeMultArabic(
                     seekBar.getContext(),
-                    ReaderTextSizeUtils.calculateMultiplier(ReaderTextSizeUtils.normalizeProgress(seekbar.getProgress()))
+                    ReaderTextSizeUtils.calculateMultiplier(
+                        ReaderTextSizeUtils.normalizeProgress(seekbar.getProgress()))
                 );
             }
         });
@@ -786,7 +761,8 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
             public void onStopTrackingTouch(@NonNull SeekBar seekBar) {
                 SPReader.setSavedTextSizeMultTransl(
                     seekBar.getContext(),
-                    ReaderTextSizeUtils.calculateMultiplier(ReaderTextSizeUtils.normalizeProgress(seekbar.getProgress()))
+                    ReaderTextSizeUtils.calculateMultiplier(
+                        ReaderTextSizeUtils.normalizeProgress(seekbar.getProgress()))
                 );
             }
         });
@@ -827,7 +803,63 @@ public class FragSettingsMain extends FragSettingsBase implements FragmentResult
     }
 
     private void iniOtherSettings() {
+        initDownloadSource(mBinding.otherSettings.getRoot());
         initLogLauncher(mBinding.otherSettings.getRoot());
+    }
+
+    private void initDownloadSource(LinearLayout parent) {
+        LytReaderSettingsItemBinding binding = LytReaderSettingsItemBinding.inflate(mInflater);
+        setupLauncherParams(R.drawable.dr_icon_download, binding);
+
+        setupDownloadSrcTitle(binding);
+        binding.launcher.setOnClickListener(v -> launchDownloadSourceSelector(v.getContext(), binding));
+
+        parent.addView(binding.getRoot());
+    }
+
+    private void launchDownloadSourceSelector(Context ctx, LytReaderSettingsItemBinding parentBinding) {
+        PeaceBottomSheet sheetDialog = new PeaceBottomSheet();
+        PeaceBottomSheetParams params = sheetDialog.getParams();
+        params.setHeaderTitle(getString(R.string.titleResourceDownloadSource));
+
+        LytResDownloadSourceSheetBinding binding = LytResDownloadSourceSheetBinding.inflate(mInflater);
+
+        binding.downloadSrcGroup.check(
+            (DOWNLOAD_SRC_GITHUB.equals(SPAppConfigs.getResourceDownloadSrc(ctx)))
+                ? R.id.srcGithub
+                : R.id.srcJsDelivr
+        );
+
+        params.setContentView(binding.getRoot());
+
+        binding.downloadSrcGroup.setOnCheckChangedListener((button, checkedId) -> {
+            final String downloadSrc;
+
+            if (checkedId == R.id.srcGithub) {
+                downloadSrc = DOWNLOAD_SRC_GITHUB;
+            } else if (checkedId == R.id.srcJsDelivr) {
+                downloadSrc = DOWNLOAD_SRC_JSDELIVR;
+            } else {
+                downloadSrc = DOWNLOAD_SRC_DEFAULT;
+            }
+
+            SPAppConfigs.setResourceDownloadSrc(ctx, downloadSrc);
+            DownloadSourceUtils.resetDownloadSourceBaseUrl(ctx);
+
+            sheetDialog.dismiss();
+
+            return Unit.INSTANCE;
+        });
+
+        sheetDialog.setOnDismissListener(() -> setupDownloadSrcTitle(parentBinding));
+        sheetDialog.show(getParentFragmentManager());
+    }
+
+    private void setupDownloadSrcTitle(LytReaderSettingsItemBinding binding) {
+        final String selectedSource = DownloadSourceUtils.getCurrentSourceName(
+            binding.getRoot().getContext()
+        );
+        prepareTitle(binding, R.string.titleResourceDownloadSource, selectedSource);
     }
 
     private void initLogLauncher(LinearLayout parent) {

--- a/app/src/main/java/com/quranapp/android/utils/app/DownloadSourceUtils.kt
+++ b/app/src/main/java/com/quranapp/android/utils/app/DownloadSourceUtils.kt
@@ -1,0 +1,48 @@
+package com.quranapp.android.utils.app
+
+import android.content.Context
+import com.quranapp.android.api.ApiConfig
+import com.quranapp.android.api.RetrofitInstance
+import com.quranapp.android.utils.sharedPrefs.SPAppConfigs
+
+object DownloadSourceUtils {
+    const val DOWNLOAD_SRC_GITHUB = "github"
+    const val DOWNLOAD_SRC_JSDELIVR = "jsdelivr"
+    const val DOWNLOAD_SRC_DEFAULT = DOWNLOAD_SRC_JSDELIVR
+
+    @JvmStatic
+    fun getCurrentSourceName(context: Context): String {
+        return when (SPAppConfigs.getResourceDownloadSrc(context)) {
+            DOWNLOAD_SRC_GITHUB -> "raw.githubusercontent.com"
+            DOWNLOAD_SRC_JSDELIVR -> "cdn.jsdelivr.net"
+            else -> "cdn.jsdelivr.net"
+        }
+    }
+
+    @JvmStatic
+    fun getDownloadSourceBaseUrl(context: Context): String {
+        return when (SPAppConfigs.getResourceDownloadSrc(context)) {
+            DOWNLOAD_SRC_GITHUB -> ApiConfig.GITHUB_ROOT_URL
+            DOWNLOAD_SRC_JSDELIVR -> ApiConfig.JS_DELIVR_ROOT_URL
+            else -> ApiConfig.JS_DELIVR_ROOT_URL
+        }
+    }
+
+    @JvmStatic
+    fun resetDownloadSourceBaseUrl(ctx: Context) {
+        val downloadSrcUrl = getDownloadSourceBaseUrl(ctx)
+
+        if (RetrofitInstance.githubResDownloadUrl != downloadSrcUrl) {
+            RetrofitInstance.githubResDownloadUrl = downloadSrcUrl
+            RetrofitInstance.resetGithubApi()
+        }
+    }
+
+    fun changeDownloadSource(ctx: Context) {
+        val currentSrc = SPAppConfigs.getResourceDownloadSrc(ctx)
+        val newSrc = if (currentSrc == DOWNLOAD_SRC_GITHUB) DOWNLOAD_SRC_JSDELIVR else DOWNLOAD_SRC_GITHUB
+
+        SPAppConfigs.setResourceDownloadSrc(ctx, newSrc)
+        resetDownloadSourceBaseUrl(ctx)
+    }
+}

--- a/app/src/main/java/com/quranapp/android/utils/sharedPrefs/SPAppConfigs.kt
+++ b/app/src/main/java/com/quranapp/android/utils/sharedPrefs/SPAppConfigs.kt
@@ -2,6 +2,7 @@ package com.quranapp.android.utils.sharedPrefs
 
 import android.annotation.SuppressLint
 import android.content.Context
+import com.quranapp.android.utils.app.DownloadSourceUtils.DOWNLOAD_SRC_DEFAULT
 
 object SPAppConfigs {
     const val SP_APP_CONFIGS = "sp_app_configs"
@@ -12,11 +13,13 @@ object SPAppConfigs {
     private const val KEY_RECITATIONS_VERSION = "key.versions.recitations"
     private const val KEY_RECITATION_TRANSLATIONS_VERSION = "key.versions.recitation_translations"
     private const val KEY_TAFSIRS_VERSION = "key.versions.tafsirs"
+    private const val KEY_RESOURCE_DOWNLOAD_SRC = "key.resource.download_src"
 
     const val LOCALE_DEFAULT = "default"
     const val THEME_MODE_DEFAULT = "app.theme.default"
     const val THEME_MODE_LIGHT = "app.theme.light"
     const val THEME_MODE_DARK = "app.theme.dark"
+
 
     private fun sp(context: Context) = context.getSharedPreferences(
         SP_APP_CONFIGS,
@@ -87,6 +90,23 @@ object SPAppConfigs {
         sp(ctx).edit().apply {
             putLong(KEY_TAFSIRS_VERSION, version)
             apply()
+        }
+    }
+
+    @JvmStatic
+    fun getResourceDownloadSrc(ctx: Context): String {
+        return sp(ctx).getString(
+            KEY_RESOURCE_DOWNLOAD_SRC,
+            DOWNLOAD_SRC_DEFAULT
+        ) ?: DOWNLOAD_SRC_DEFAULT
+    }
+
+    @SuppressLint("ApplySharedPref")
+    @JvmStatic
+    fun setResourceDownloadSrc(ctx: Context, src: String?) {
+        sp(ctx).edit().apply {
+            putString(KEY_RESOURCE_DOWNLOAD_SRC, src)
+            commit()
         }
     }
 }

--- a/app/src/main/res/layout/lyt_res_download_source_sheet.xml
+++ b/app/src/main/res/layout/lyt_res_download_source_sheet.xml
@@ -1,0 +1,58 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.core.widget.NestedScrollView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:overScrollMode="ifContentScrolls">
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="vertical"
+        android:padding="10dp">
+
+        <androidx.appcompat.widget.AppCompatTextView
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginBottom="10dp"
+            android:gravity="center"
+            android:text="@string/msgResourceDownloadSource"
+            android:textColor="@color/colorText2"
+            android:textSize="@dimen/dmnCommonSize2" />
+
+        <com.quranapp.android.widgets.radio.PeaceRadioGroup
+            android:id="@+id/downloadSrcGroup"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
+
+            <com.quranapp.android.widgets.radio.PeaceRadioButton
+                android:id="@id/srcGithub"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_margin="2dp"
+                android:background="@drawable/dr_bg_hover_cornered"
+                android:gravity="center"
+                android:padding="10dp"
+                android:text="GitHub"
+                android:textAppearance="@style/TextAppearanceCommonTitle"
+                app:peaceComp_spaceBetween="15dp"
+                app:peaceComp_subText="raw.githubusercontent.com"
+                tools:ignore="HardcodedText" />
+
+            <com.quranapp.android.widgets.radio.PeaceRadioButton
+                android:id="@id/srcJsDelivr"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_margin="2dp"
+                android:background="@drawable/dr_bg_hover_cornered"
+                android:gravity="center"
+                android:padding="10dp"
+                android:text="JSDelivr"
+                android:textAppearance="@style/TextAppearanceCommonTitle"
+                app:peaceComp_spaceBetween="15dp"
+                app:peaceComp_subText="cdn.jsdelivr.net"
+                tools:ignore="HardcodedText" />
+        </com.quranapp.android.widgets.radio.PeaceRadioGroup>
+    </LinearLayout>
+</androidx.core.widget.NestedScrollView>

--- a/app/src/main/res/values/ids.xml
+++ b/app/src/main/res/values/ids.xml
@@ -12,6 +12,9 @@
     <item name="on" type="id" />
     <item name="off" type="id" />
 
+    <item name="srcGithub" type="id" />
+    <item name="srcJsDelivr" type="id" />
+
     <item name="btnPlayControl" type="id" />
     <item name="btnFootnotes" type="id" />
     <item name="btnTafsir" type="id" />

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -407,4 +407,6 @@
     <string name="strPreviewTextTransl">In the Name of Allâh, the Most Gracious, the Most Merciful</string>
     <string name="msgClearReadHistory">Clear read history</string>
     <string name="msgClearSearchHistory">Clear search history</string>
+    <string name="titleResourceDownloadSource">Resource download source</string>
+    <string name="msgResourceDownloadSource">Use this option to download Quran data from a different source if one is not working</string>
 </resources>

